addons/web_tour/static/src/js/tip.js

```diff
@@ -113,16 +124,100 @@ var Tip = Widget.extend({
         if (!$anchor.is(this.$anchor)) {
             this._unbind_anchor_events();
             this.$anchor = $anchor;
-            this.$ideal_location = this._get_ideal_location();
             if (this.$el.parent()[0] !== document.body) {
-                this.$el.appendTo(this.$ideal_location);
+                if (this.viewPortState === 'in') {
+                    this.$actualAnchor = $anchor;
+                    this.$ideal_location = this._get_ideal_location($anchor);
+                    this.$furtherIdealLocation = this._get_ideal_location(this.$ideal_location);
+                    this.$el.appendTo(this.$ideal_location);
+                } else {
+                    this.$ideal_location = this._get_ideal_location($anchor);
+                    this.$furtherIdealLocation = this._get_ideal_location(this.$ideal_location);
+                    this.$actualAnchor = this.$ideal_location;
+                    this.$el.appendTo(this.$furtherIdealLocation);
+                }
             }
             this._bind_anchor_events();
         }
         this._reposition();
     },
-    _get_ideal_location: function () {
-        var $location = this.$anchor;
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * This function figure out which direction the tip should take
+     * and if he's at the bottom or the top of the targeted element
+     * or if it's an indicator to scroll.
+     *
+     * If the direction or the anchor change, then we adapth the frontend
+     * position of the tip.
+     */
+    _update_position: function () {
+        this.moveTip = false;
+        let halfHeight = 0;
+        if (this.initialPosition === 'right' || this.initialPosition === 'left') {
+            halfHeight = this.$anchor.innerHeight() / 2;
+        }
+        let position = this.info.position;
+        /* Check if the viewport state change to know if we need to move the anchor of the tip.
+        * up : the target element is above the current viewport
+        * down : the target element is below the current viewport
+        * in : the target element is in the current viewport
+        */
+        let viewPortState = "in";
+        const paddingTop = this.$ideal_location.css('padding-top') ? parseInt(this.$ideal_location.css('padding-top')) : 0;
```

Comment:
You read `this.$ideal_location.css('padding-top')` twice here, use a variable then parse if necessary.